presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/TimestampRepresentations.java

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.base.type;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.Int96ArrayBlock;
+import io.prestosql.spi.block.LongArrayBlock;
+import io.prestosql.spi.type.LongTimestamp;
+import io.prestosql.spi.type.TimestampType;
+import org.joda.time.DateTimeZone;
+
+import static io.prestosql.spi.type.TimestampType.MAX_SHORT_PRECISION;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_NANOS;
+import static io.prestosql.spi.type.Timestamps.MICROSECONDS_PER_MILLISECOND;
+import static io.prestosql.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
+import static io.prestosql.spi.type.Timestamps.PICOSECONDS_PER_MICROSECOND;
+import static io.prestosql.spi.type.Timestamps.epochMicrosToMillisWithRounding;
+import static io.prestosql.spi.type.Timestamps.round;
+import static io.prestosql.spi.type.Timestamps.roundDiv;
+import static java.lang.Math.multiplyExact;
+import static java.util.Objects.requireNonNull;
+
+public final class TimestampRepresentations
+{
+    private TimestampRepresentations() {}
+
+    public static PrestoTimestampEncoder<? extends Comparable<?>> createTimestampEncoder(TimestampType type, DateTimeZone timeZone)
+    {
+        requireNonNull(type, "type is null");
+        requireNonNull(timeZone, "timeZoneKey is null");
+
+        if (type.isShort()) {
+            return new ShortTimestampEncoder(type, timeZone);
+        }
+        return new LongTimestampEncoder(type, timeZone);
+    }
+
+    // copied from io.prestosql.type.DateTimes
+    static LongTimestamp longTimestamp(long epochSecond, long fractionInPicos)
+    {
+        return new LongTimestamp(
+                multiplyExact(epochSecond, MICROSECONDS_PER_SECOND) + fractionInPicos / PICOSECONDS_PER_MICROSECOND,
+                (int) (fractionInPicos % PICOSECONDS_PER_MICROSECOND));
+    }
+
+    /**
+     * Provide a canonical representation of the timestamp at the given position in the block with the precision
+     * of the target type.  Uses {@link LongTimestamp}, since it can represent any valid precision.  This is
+     * effectively a cast to a different precision timestamp and is useful in CTAS, when the engine won't do the
+     * cast for us.
+     */
+    public static LongTimestamp longTimestampRepresentation(Block block, int position, TimestampType type)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        if (block instanceof LongArrayBlock) {
+            long epochMicros = TIMESTAMP_MICROS.getLong(block, position);
+            if (type == TIMESTAMP_MILLIS) {
+                epochMicros = epochMicrosToMillisWithRounding(epochMicros) * MICROSECONDS_PER_MILLISECOND;
```

Comment:
if the `type == TIMESTAMP_MILLIS` the data coming in should have zeroed out micros part, so the rounding should not be necessary. Is it necessary due to CTAS flow? At least a comment, but we should talk this through.

Reply:
See comment above.